ngsPETSc/ts.py

```diff
@@ -0,0 +1,203 @@
+        self.second_order = False
+        self.trial = info[0]
+        self.t = info[1]
+        self.F = F
+        self.G = G
+        if "ngs_jacobian_mat_type" not in solverParameters:
+            solverParameters["ngs_jacobian_mat_type"] = "aij"
+        jacobianMatType = solverParameters["ngs_jacobian_mat_type"]
+        self.ts = PETSc.TS().create(comm=dofs.comm.mpi4py)
+        #Deafult TS setup
+        self.ts.setExactFinalTime(3)
+        self.ts.setMaxSNESFailures(-1)
+        #Setting up the options
+        options_object = PETSc.Options()
+        if solverParameters is not None:
+            for optName, optValue in solverParameters.items():
+                options_object[optName] = optValue
+        self.ts.setOptionsPrefix(optionsPrefix)
+        self.ts.setFromOptions()
+        #Setting up utility for mappings
+        self.vectorMapping = VectorMapping(self.fes)
+        if G is not None:
+            def rhs(t, x): #pylint: disable=E0102, E0213
+                self.t.Set(t)
+                f = GridFunction(self.fes)
+                a = BilinearForm(self.fes)
+                a += self.G
+                a.Apply(x.vec, f.vec)
+                return f
+            self.rhs = rhs
+        if F is not None:
+            def residual(t, x, xdot): #pylint: disable=E0102, E0213
+                res = GridFunction(self.fes)
+                self.t.Set(t)
+                updatedF = self.F.Replace({self.trial.dt: xdot})
+                a = BilinearForm(self.fes)
+                a += updatedF
+                a.Apply(x.vec, res.vec)
+                return  res
+            self.residual = residual
+        else:
+            raise ValueError("You need to provide a residual function or a variational form F.")
+        if jacobian is not None:
+            self.jacobian = jacobian
+            self.jacobianMatType = jacobianMatType
+            self.second_order = True
+        elif F is not None:
+            def jacobian(x, t, xdot): #pylint: disable=E0102,E0213
+                self.t.Set(t)
+                updatedF = self.F.Replace({self.trial.dt: xdot})
+                a = BilinearForm(self.fes)
+                a += updatedF
+                a.AssembleLinearization(x.vec)
+                return a.mat
+            self.jacobian = jacobian
+            self.second_order = False
+            self.jacobianMatType = jacobianMatType
+        else:
+            raise ValueError("You need to provide a jacobian function or a variational form F.")
+    def setup(self, timeInfo):
+        '''
+        This is method is used to setup the PETSc TS object
+        '''
+        if isinstance(timeInfo, (list,tuple)) and len(timeInfo) == 3:
+            self.t0, self.tf, self.dt = timeInfo
+        elif isinstance(timeInfo, dict):
+            self.t0 = timeInfo["t0"]
+            self.tf = timeInfo["tf"]
+            self.dt = timeInfo["dt"]
+        else:
+            raise ValueError("timeInfo must be a list/tuple or a dict")
```

Comment:
why timeInfo is here ?
these should be quantities decided at initialization time, not setUp.
Take a look at how I do TS customization in deal.II https://github.com/dealii/dealii/blob/master/include/deal.II/lac/petsc_ts.h#L44